mongo/single_result.go

```diff
@@ -33,6 +33,9 @@ type SingleResult struct {
 // will be returned. If there were no returned documents, ErrNoDocuments is
 // returned.
 func (sr *SingleResult) Decode(v interface{}) error {
+	if sr.err != nil {
```

Comment:
Since we check for this in the switch, can we just move the `sr.reg == nil` check into the `case sr.rdr != nil:` and `case sr.cur != nil:` cases? It'll be duplicating the three lines of code but it'll make this less complex overall.

Reply:
I just updated it. I think it makes sense to just pull the case err != nil out instead and add the else if to indicate that evaluation order is important outside of the switch.